Source/Core/Common/Network.cpp

```diff
@@ -170,6 +194,85 @@ u8 UDPHeader::IPProto() const
   return static_cast<u8>(IPPROTO_UDP);
 }
 
+ARPHeader::ARPHeader() = default;
+
+ARPHeader::ARPHeader(u32 from_ip, const MACAddress& from_mac, u32 to_ip, const MACAddress& to_mac)
+{
+  hardware_type = htons(BBA_HARDWARE_TYPE);
+  protocol_type = IPV4_HEADER_TYPE;
+  hardware_size = MAC_ADDRESS_SIZE;
+  protocol_size = IPV4_ADDR_LEN;
+  opcode = 0x200;
+  sender_ip = from_ip;
+  target_ip = to_ip;
+  targer_address = to_mac;
+  sender_address = from_mac;
+}
+
+u16 ARPHeader::Size() const
+{
+  return static_cast<u16>(SIZE);
+}
+
+DHCPBody::DHCPBody() = default;
+
+DHCPBody::DHCPBody(u32 transaction, const MACAddress& client_address, u32 new_ip, u32 serv_ip)
+{
+  transaction_id = transaction;
+  message_type = DHCPConst::MESSAGE_REPLY;
+  hardware_type = BBA_HARDWARE_TYPE;
+  hardware_addr = MAC_ADDRESS_SIZE;
+  client_mac = client_address;
+  your_ip = new_ip;
+  server_ip = serv_ip;
+}
+
+DHCPPacket::DHCPPacket() = default;
+
+DHCPPacket::DHCPPacket(const std::vector<u8>& data)
+{
+  if (data.size() < DHCPBody::SIZE)
+    return;
+  body = Common::BitCastPtr<DHCPBody>(data.data());
+  u16 offset = DHCPBody::SIZE;
+
+  while (offset < data.size())
+  {
+    const u8 fnc = data[offset];
+    const u8 len = data[offset + 1];
+    if (fnc == 255)
+      break;
+    const auto opt_begin = data.begin() + offset;
+    offset += 2 + len;
+    if (offset >= data.size())
```

Comment:
Is this right? I would assume going up to exactly the end of the data would be fine, so I would expect a `if (offset > data.size())` here.

Reply:
the data is 0 based, so size of 10 = data 0-9
so if i read at 10, im out of bound

Reply:
Well yes, but you're passing an end iterator, which is allowed to be one-past-the-end (and indeed must be if you're trying to read the last element by passing a range) as long as you don't dereference it. Though admittedly in practice this probably doesn't matter since a well-formed packet must end with a singular `0xff` byte anyway.